### PR TITLE
@ #497 | should support register with weight

### PR DESCRIPTION
**Pull request by hieptranquoc**
please check this PR, thank you!
connect #497

### CODE REVIEW

**Comment by datphan** [comment]
I think we should try this: http://ruby-doc.org/core-2.2.2/Enumerable.html#method-i-sort_by

**Comment by hieptranquoc** [comment]
@datphan I've updated, please check this PR again!

File: lib/teracy-dev/config/manager.rb [hunk 1]
```diff
@@ -8,23 +8,37 @@ class Manager
...
-      def register(configurator)
+      def register(configurator, weight)
         if !configurator.respond_to?(:configure)
           @logger.warn("configurator #{configurator} must implement configure method, ignored")
           return
         end
-        @configurators << configurator
+
+        unless weight.is_a? Integer
+          @logger.warn("Weight must be integer, otherwise weight will be set to default(5)")
```
Comment (datphan) [inline lib/teracy-dev/config/manager.rb]: should be `"#{configurator}'s weight must be integer, ..."`

**Review by datphan — DISMISSED** [review]

File: lib/teracy-dev/config/manager.rb (hunk 1 above)
Comment (hoatle) [reply lib/teracy-dev/config/manager.rb]: weight must be an integer within 0-9 range, make sure to validate that.
- not set: default 5
- wrong type or wrong value => add warn and set default 5

File: lib/teracy-dev/processors/manager.rb [hunk 2]
```diff
@@ -12,19 +12,33 @@ def initialize
         @@instance = self
 
         @logger = Logging.logger_for(self.class.name)
-        @processors = []
-
+        @processors = Array.new
```
Comment (hoatle) [inline lib/teracy-dev/processors/manager.rb]: `@processors = []` is shorter, it's the same with Array.new, there is no reason to change this

File: lib/teracy-dev/processors/manager.rb [hunk 3]
```diff
@@ -12,19 +12,33 @@ def initialize
...
-      def register(processor)
+      def register(processor, weight)
         if !processor.respond_to?(:process)
           @logger.warn("processor #{processor} must implement process method, ignored")
           return
         end
-        @processors << processor
+
+        unless weight.is_a? Integer
+          @logger.warn("Weight must be integer, otherwise weight will be set to default(5)")
```
Comment (hoatle) [inline lib/teracy-dev/processors/manager.rb]: message should show which processor has this warning (refer to @datphan's comment)

File: lib/teracy-dev/loader.rb [hunk 4]
```diff
@@ -32,6 +32,7 @@ def start
       sync()
       settings = build_settings().freeze
       require_teracy_dev_version(settings['teracy-dev']['require_version'])
+      @configManager.sort_configurator_by_weight
```
Comment (hoatle) [inline lib/teracy-dev/loader.rb]: move this logic out of loader, to the configManager instead

File: lib/teracy-dev/loader.rb [hunk 5]
```diff
@@ -100,6 +101,7 @@ def sync_teracy_dev_entry(location)
     def build_settings
       settings = @settingsManager.build_settings(@extension_entry_path)
       load_extension_entry_files(settings)
+      @processorsManager.sort_processor_by_weight
```
Comment (hoatle) [inline lib/teracy-dev/loader.rb]: move this logic to processorsManager instead

File: lib/teracy-dev/config/manager.rb [hunk 6]
```diff
@@ -8,23 +8,37 @@ class Manager
...
+        @hash_to_sort = Hash.new{ |key,val| key[val] = [] }
       end
 
-      def register(configurator)
+      def register(configurator, weight)
         if !configurator.respond_to?(:configure)
           @logger.warn("configurator #{configurator} must implement configure method, ignored")
           return
         end
-        @configurators << configurator
```
Comment (hoatle) [inline lib/teracy-dev/config/manager.rb:24]: I prefer this to store the original registered order
```
@configurators << [configurator, weight]
```

Then from the process method, we sort it by the specified requirements and call each configurator, this should make sure the logic is clear, we could change the sort requirement later without having changing the logic in mind, just update the sort function.

**Review by hoatle — CHANGES_REQUESTED** [review]
@hieptranquoc pls check my comments

**Comment by hoatle** [comment]
@datphan not sure why you approved this when this is not good enough.

File: lib/teracy-dev/config/manager.rb (hunk 6 above)
Comment (hoatle) [reply lib/teracy-dev/config/manager.rb:24]: this helps to debug a lot, too.

**Comment by hieptranquoc** [comment]
@hoatle please check again! I've updated.

File: lib/teracy-dev/config/manager.rb [hunk 7]
```diff
@@ -8,25 +8,48 @@ class Manager
...
+            return
+          end
+        end
+
+        hash_to_sort = Hash.new{ |key,val| key[val] = [] }
+        @configurators.each { |index| hash_to_sort[index[:weight]].push(index[:configurator]) }
+        sorted = hash_to_sort.sort_by { |key,val| key }.reverse
+        @configurators.clear
+        sorted.each do |key,val|
+          @configurators.concat(val)
```
Comment (hoatle) [inline lib/teracy-dev/config/manager.rb]: don't change the original configurators

File: lib/teracy-dev/config/manager.rb [hunk 8]
```diff
@@ -8,25 +8,48 @@ class Manager
...
           return
         end
-        @configurators << configurator
+
+        unless weight.is_a? Integer and (0..9).include?(weight)
+          @logger.warn("#{configurator}'s weight must be integer and have value in range 0.. 9, otherwise weight will be set to default (5)")
+          weight = 5
+        end
+
+        @configurators << { configurator: configurator, weight: weight }
```
Comment (hoatle) [inline lib/teracy-dev/config/manager.rb]: try this: https://stackoverflow.com/questions/28234803/ruby-sort-by-multiple-fields-and-multilple-directions-for-different-data-types

processors << [id: processors.length, processor: processor, weight: weight]

and to sort with our requirement:

sorted_processors = multi_sort(processors, id: :asc, weight: :desc)

File: lib/teracy-dev/processors/manager.rb [hunk 9]
```diff
@@ -13,20 +13,43 @@ def initialize
...
+          @processors.concat(val)
+        end
+      end
+
       # run the pipelines
       def process(settings)
+        sort_processor_by_weight()
         @logger.debug("start processing: #{settings}")
 
         @processors.each do |processor|
```
Comment (hoatle) [inline lib/teracy-dev/processors/manager.rb]: `Util.multi_sort(@processors, id: :asc, weight: :desc).each do |processor|`

**Comment by hieptranquoc** [comment]
@hoatle I've updated! Please check again!

**Review by hoatle — APPROVED** [review]
LGTM, merging and I'll do some clean up

**Pull request merged**